app/src/main/java/bg/zahov/app/ui/workout/WorkoutViewModelNew.kt

```diff
@@ -0,0 +1,905 @@
+                    _uiState.update { old ->
+                        val exercisesToUpdate = _uiState.value.exercises.toMutableList()
+                        exercisesToUpdate.addAll(it.toWorkoutEntryList(templateExercises))
+                        old.copy(exercises = exercisesToUpdate)
+                    }
+                    addExerciseToWorkoutProvider.resetSelectedExercises()
+                }
+            }
+            // Collecting the rest timer
+            launch {
+                restTimerProvider.restTimer.collect {
+                    it.elapsedTime?.let { time ->
+                        _uiState.update { old ->
+                            old.copy(restTimer = time)
+                        }
+                    }
+                }
+            }
+            // Collecting rest state
+            launch {
+                restTimerProvider.restState.collect {
+                    _uiState.update { old ->
+                        old.copy(restState = it)
+                    }
+                }
+            }
+            // collecting template exercises
+            launch {
+                repo.getTemplateExercises().collect {
+                    templateExercises = it
+                }
+            }
+        }
+    }
+
+    /**
+     * A way to know which exercise to replace once we come back
+     * with a selected exercise to replace
+     *
+     * @param itemPosition The index of the exercise to replace
+     */
+    fun replaceExercise(itemPosition: Int) {
+        exerciseToReplaceIndex = itemPosition
+    }
+
+    /**
+     * Removes the exercise for the current position
+     * then iterates to remove its sets if there are any
+     *
+     * @param position The index of the exercise to remove
+     */
+    fun removeExercise(position: Int) {
+        _uiState.update { old ->
+            val newExercises = old.exercises.toMutableList()
+
+            newExercises.removeAt(position)
+
+            while (position < newExercises.size && newExercises[position] is WorkoutEntry.SetEntry) {
+                newExercises.removeAt(position)
+            }
+
+            old.copy(exercises = newExercises)
+        }
+    }
+
+    /**
+     * Adds a new set to the workout at the specified position.
+     *
+     * @param position The index of the exercise in the workout list to which the set should be added.
+     *
+     * This function determines the correct position to insert the new set based on the workout structure
+     * and ensures that the set is associated with the appropriate exercise. It handles edge cases, such as when
+     * the exercise is the last entry in the list or when there is only one exercise in the workout. The new set
+     * is configured using data from the corresponding template exercise if available.
+     */
+    fun addSet(position: Int) {
+        _uiState.update { old ->
+
+            var edgeCaseFlag = false
```

Comment:
I don't think this flag has a reliable name, try to name it more explicit, or even remove it if your logic seems incorrect